bindings/rust/evmc-vm/src/lib.rs

```diff
@@ -2,6 +2,195 @@ pub extern crate evmc_sys;
+        self.status_code
+    }
+
+    pub fn get_gas_left(&self) -> i64 {
+        self.gas_left
+    }
+
+    pub fn get_output(&self) -> Option<&Vec<u8>> {
+        self.output.as_ref()
+    }
+
+    pub fn get_create_address(&self) -> Option<&ffi::evmc_address> {
+        // Only return Some if the address is valid (e.g. the status is EVMC_SUCCESS)
+        if self.status_code == ffi::evmc_status_code::EVMC_SUCCESS {
+            Some(&self.create_address)
+        } else {
+            None
+        }
+    }
+}
+
+impl From<ffi::evmc_result> for ExecutionResult {
+    fn from(result: ffi::evmc_result) -> Self {
+        let ret = ExecutionResult {
+            status_code: result.status_code,
+            gas_left: result.gas_left,
+            output: if !result.output_data.is_null() {
+                // Pre-allocate a vector.
+                let mut buf: Vec<u8> = Vec::with_capacity(result.output_size);
+
+                unsafe {
+                    // Set the len of the vec manually.
+                    buf.set_len(result.output_size);
+                    // Copy from the C struct's buffer to the vec's buffer.
+                    std::ptr::copy(result.output_data, buf.as_mut_ptr(), result.output_size);
+                }
+
+                Some(buf)
+            } else {
+                None
+            },
+            create_address: result.create_address,
+        };
+
+        // Release allocated ffi struct.
+        unsafe {
+            if result.release.is_some() {
+                result.release.unwrap()(&result as *const ffi::evmc_result);
+            }
+        }
+
+        ret
+    }
+}
+
+/// Returns a pointer to a heap-allocated evmc_result.
+impl Into<*const ffi::evmc_result> for ExecutionResult {
+    fn into(self) -> *const ffi::evmc_result {
+        let (buffer, len) = if let Some(buf) = self.output {
+            let buf_len = buf.len();
+
+            // Manually allocate heap memory for the new home of the output buffer.
+            let memlayout = std::alloc::Layout::from_size_align(buf_len, 1).expect("Bad layout");
+            let new_buf = unsafe { std::alloc::alloc(memlayout) };
+            unsafe {
+                // Copy the data into the allocated buffer.
+                std::ptr::copy(buf.as_ptr(), new_buf, buf_len);
+            }
+
+            (new_buf as *const u8, buf_len)
+        } else {
+            (std::ptr::null(), 0)
+        };
+
+        Box::into_raw(Box::new(ffi::evmc_result {
+            status_code: self.status_code,
+            gas_left: self.gas_left,
+            output_data: buffer,
+            output_size: len,
+            release: Some(release_result),
+            create_address: self.create_address,
+            padding: [0u8; 4],
+        }))
+    }
+}
+
+/// Callback to pass across FFI, de-allocating the struct.
+extern "C" fn release_result(result: *const ffi::evmc_result) {
+    unsafe {
+        if !(*result).output_data.is_null() {
```

Comment:
Actually another question: would the code be cleaner if you do `Box::from_raw` first and call dealloc on the boxed object?